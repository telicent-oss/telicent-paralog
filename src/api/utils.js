import config from "config/app-config";

const createParalogEndpoint = (path) => `${config.api.url}/${path}`;
const createOntologyServiceEndpoint = (path) =>
  `${config.services.ontology}/${path}`;

const fetchOptions = {
  headers: {
    "Content-Type": "application/json",
  },
};

export { createParalogEndpoint, createOntologyServiceEndpoint, fetchOptions };
