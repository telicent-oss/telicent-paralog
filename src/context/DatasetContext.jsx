import { createContext, useState } from "react";

const DatasetContext = createContext();

const DatasetProvider = ({ children }) => {
  const [selectedFloodAreas, setSelectedFloodAreas] = useState([]);
  const [expandedFloodAreas, setExpandedFloodAreas] = useState([]);
  const [selectedAssets, setSelectedAssets] = useState([]);
  const [expandedAssets, setExpandedAssets] = useState([]);

  const addSelectedFloodAreas = (polygonUris) => {
    setSelectedFloodAreas(polygonUris);
  };

  const addSelectedAssets = (assets) => {
    console.log({ assets });
    setSelectedAssets(assets);
  };

  const removeSelectedAsset = (selectedAsset) => {
    setSelectedAssets((assets) => assets.filter((asset) => asset !== selectedAsset));
  };

  const toggleExpandedAssets = (selectedGroup) => {
    const index = expandedAssets.findIndex((ontologyGroup) => ontologyGroup === selectedGroup);
    if (index === -1) {
      setExpandedAssets([...expandedAssets, selectedGroup]);
      return;
    }
    const filteredGroups = expandedAssets.filter(
      (ontologyGroup) => ontologyGroup !== selectedGroup
    );
    setExpandedAssets(filteredGroups);
  };

  const updateExpandedFloodAreas = (groups) => {
    setExpandedFloodAreas(groups);
  };

  return (
    <DatasetContext.Provider
      value={{
        expandedAssets,
        expandedFloodAreas,
        selectedFloodAreas,
        selectedAssets,
        addSelectedAssets,
        addSelectedFloodAreas,
        removeSelectedAsset,
        toggleExpandedAssets,
        updateExpandedFloodAreas,
      }}
    >
      {children}
    </DatasetContext.Provider>
  );
};

export { DatasetContext, DatasetProvider };
