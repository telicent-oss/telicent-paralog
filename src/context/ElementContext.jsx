import React, { useCallback, useReducer, useState } from "react";

import elementsReducer, {
  ADD_ASSETS,
  ADD_DEPENDENCIES,
  AREA_SELECTION,
  CLEAR_SELECTED,
  DISMISS_ERROR,
  FILTER_SELECTED_ELEMENTS,
  INITIAL_STATE,
  MULTI_SELECT_ELEMENTS,
  REMOVE_ELEMENTS_BY_TYPE,
  RESET,
  SELECT_ELEMENT,
  UPDATE_ERRORS,
} from "./elements-reducer";

export const ElementsContext = React.createContext();

export const ElementsProvider = ({ children }) => {
  const [state, dispatch] = useReducer(elementsReducer, INITIAL_STATE);
  const [selectedTimeline, setSelectedTimeline] = useState(null);

  const onFloodTimelineSelect = (selected) => {
    if (selected !== selectedTimeline) {
      setSelectedTimeline(selected);
    }
    return;
  };

  const closeTimelinePanel = () => {
    setSelectedTimeline(null);
  };

  const {
    assets,
    dependencies,
    errors,
    selectedElements,
    maxAssetCriticality,
    maxAssetTotalCxns,
    assetCriticalityColorScale,
    cxnCriticalityColorScale,
    totalCxnsColorScale,
  } = state;

  const filterSelectedElements = useCallback(() => {
    dispatch({ type: FILTER_SELECTED_ELEMENTS });
  }, []);

  const addElements = useCallback(
    (assets, dependencies) => {
      if (!Array.isArray(assets) || !Array.isArray(dependencies)) return;
      dispatch({ type: ADD_ASSETS, assets });
      dispatch({ type: ADD_DEPENDENCIES, dependencies });
      filterSelectedElements();
    },
    [filterSelectedElements]
  );

  const removeElementsByType = useCallback(
    (typeUri) => {
      if (!typeUri) return;
      dispatch({ type: REMOVE_ELEMENTS_BY_TYPE, typeUri });
      filterSelectedElements();
    },
    [filterSelectedElements]
  );

  const reset = useCallback(() => {
    dispatch({ type: RESET });
  }, []);

  const onElementClick = useCallback((multiSelect, selectedElements) => {
    if (multiSelect) {
      dispatch({ type: MULTI_SELECT_ELEMENTS, selectedElements });
      return;
    }
    dispatch({ type: SELECT_ELEMENT, selectedElements });
  }, []);

  const onAreaSelect = useCallback((selectedElements) => {
    if (!Array.isArray(selectedElements)) return;
    dispatch({ type: AREA_SELECTION, selectedElements });
  }, []);

  const updateErrorNotifications = useCallback((msg) => {
    dispatch({ type: UPDATE_ERRORS, error: msg });
  }, []);

  const dismissErrorNotification = (error) => {
    dispatch({ type: DISMISS_ERROR, error });
  };

  const clearSelectedElements = useCallback(() => {
    dispatch({ type: CLEAR_SELECTED });
  }, []);

  return (
    <ElementsContext.Provider
      value={{
        assets,
        dependencies,
        errors,
        selectedElements,
        maxAssetCriticality,
        maxAssetTotalCxns,
        assetCriticalityColorScale,
        cxnCriticalityColorScale,
        totalCxnsColorScale,
        addElements,
        clearSelectedElements,
        dismissErrorNotification,
        filterSelectedElements,
        onAreaSelect,
        onElementClick,
        reset,
        removeElementsByType,
        updateErrorNotifications,
        selectedTimeline,
        closeTimelinePanel,
        onFloodTimelineSelect,
      }}
    >
      {children}
    </ElementsContext.Provider>
  );
};
