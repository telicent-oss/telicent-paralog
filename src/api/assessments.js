import { createParalogEndpoint, fetchOptions } from "./utils";

export const fetchAssessments = async () => {
  const response = await fetch(
    createParalogEndpoint("assessments"),
    fetchOptions,
  );

  if (!response.ok) {
    throw new Error(`Failed to retrieve assessments`);
  }

  return response.json();
};

export const fetchAssetTypes = async (assessment) => {
  const queryParams = new URLSearchParams({ assessment }).toString();
  const response = await fetch(
    createParalogEndpoint(`assessments/asset-types?${queryParams}`),
    fetchOptions,
  );

  if (!response.ok) {
    throw new Error(
      `An error occurred while retrieving data types for ${assessment}`,
    );
  }

  return response.json();
};

export const fetchAssessmentAssets = async (assessment, types) => {
  const typeParams = types.map((type) => ["types", type]);
  const queryParams = new URLSearchParams([
    ["assessment", assessment],
    ...typeParams,
  ]).toString();

  const response = await fetch(
    createParalogEndpoint(`assessments/assets?${queryParams}`),
    fetchOptions,
  );

  if (!response.ok) {
    throw new Error(
      `An error occurred while retrieving assets for assessment ${assessment} and types ${types.toString()}`,
    );
  }

  return response.json();
};

export const fetchAssessmentDependencies = async (assessment, types) => {
  const typeParams = types.map((type) => ["types", type]);
  const queryParams = new URLSearchParams([
    ["assessment", assessment],
    ...typeParams,
  ]).toString();

  const response = await fetch(
    createParalogEndpoint(`assessments/dependencies?${queryParams}`),
    fetchOptions,
  );

  if (!response.ok) {
    throw new Error(
      `An error occurred while retrieving dependencies for assessment ${assessment} and types ${typeParams.toString()}`,
    );
  }

  return response.json();
};
