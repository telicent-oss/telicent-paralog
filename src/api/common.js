import { createParalogEndpoint, fetchOptions } from "./utils";

export const fetchTypeSuperclass = async (typeUri) => {
  const queryParams = new URLSearchParams({ classUri: typeUri }).toString();
  const response = await fetch(
    createParalogEndpoint(`ontology/class?${queryParams}`),
    fetchOptions,
  );
  if (!response.ok) {
    return {};
  }
  return response.json();
};

export const fetchResidentialInformation = async (personUri) => {
  const queryParams = new URLSearchParams({ personUri }).toString();
  const response = await fetch(
    createParalogEndpoint(`person/residences?${queryParams}`),
    fetchOptions,
  );
  if (!response.ok) {
    throw new Error(
      "An error occurred while retrieving residential information",
    );
  }
  return response.json();
};

export const fetchFloodTimeline = async (floodArea) => {
  const queryParam = new URLSearchParams({
    parent_uri: `http://environment.data.gov.uk/flood-monitoring/id/floodAreas/${floodArea}`,
  }).toString();
  const response = await fetch(
    createParalogEndpoint(`states?${queryParam}`),
    fetchOptions,
  );

  if (!response.ok) {
    throw new Error(
      `An error occurred while retrieving flood timeline for Flood Area ${floodArea}`,
    );
  }

  return response.json();
};

export const fetchFloodMonitoringStations = async () => {
  const response = await fetch(
    "https://environment.data.gov.uk/flood-monitoring/id/stations?catchmentName=Isle%20of%20Wight",
  );

  if (!response.ok) {
    throw new Error(
      "An error occurred while retrieving flood monitoring stations for the Kangaroo Island",
    );
  }
  return response.json();
};

export const fetchBuildingsEpcRating = async () => {
  const response = await fetch(createParalogEndpoint("buildings"));

  if (!response.ok) {
    throw new Error(
      "An error occurred while retrieving building epc ratings for the Kangaroo Island",
    );
  }
  return response.json();
};
