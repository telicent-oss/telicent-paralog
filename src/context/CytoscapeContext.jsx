import { isEmpty } from "lodash";
import React, { createContext, useCallback, useRef } from "react";

import { useLocalStorage } from "hooks";
import { getUniqueElements } from "utils";

export const CytoscapeContext = createContext();

export const CytoscapeProvider = ({ children }) => {
  const cyRef = useRef({});
  const [layout, setLayout] = useLocalStorage("graphLayout", "cola");

  const moveTo = ({ areaSelect, cachedElements, selectedElements }) => {
    if (!cyRef.current) return;
    const padding = 20;
    const selected = areaSelect
      ? selectedElements
      : getUniqueElements([...cachedElements, ...selectedElements]);

    const elements = cyRef.current.elements().filter((element) => {
      return selected.some((selectedElement) => {
        const data = element.data("element");
        return selectedElement.uri === data.uri;
      });
    });

    fit(elements, padding);
  };

  const fit = (elements, padding) => {
    if (!cyRef.current) return;
    cyRef.current.fit(elements, padding);
  };

  const updateLayout = (layout) => {
    setLayout(layout);
  };

  const resize = () => {
    if (isEmpty(cyRef?.current)) return;
    cyRef.current.resize();
  };

  const runLayout = useCallback(() => {
    if (isEmpty(cyRef?.current) || cyRef.current?.destroyed()) return;
    const cylayout = cyRef.current.layout({ name: layout });
    cylayout.run();
  }, [cyRef, layout]);

  return (
    <CytoscapeContext.Provider
      value={{ cyRef, layout, fit, moveTo, resize, runLayout, updateLayout }}
    >
      {children}
    </CytoscapeContext.Provider>
  );
};
