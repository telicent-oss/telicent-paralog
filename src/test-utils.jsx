import { render, screen } from "@testing-library/react";
import React from "react";
import userEvent from "@testing-library/user-event";
import { Provider as UseHttpProvider } from "use-http";
import { MapProvider } from "react-map-gl";
import { QueryClient, QueryClientProvider } from "@tanstack/react-query";
import { DSProviders } from "@telicent-oss/ds";
import OntologyService from "@telicent-oss/ontologyservice";

import { createAssets, createDependencies } from "components/Dataset/dataset-utils";
import config from "config/app-config";
import { Dataset } from "./components";
import { CytoscapeProvider, ElementsContext, ElementsProvider } from "./context";

const user = userEvent.setup();

const ontologyService = new OntologyService(config.services.ontology, "/ontology");

export const clickSmallWindFarm = async () => {
  await user.click(screen.getByRole("button", { name: /Green grid/i }));
  await user.click(await screen.findByRole("checkbox", { name: "Energy [25]" }));
  expect(screen.getByRole("checkbox", { name: "Energy [25]" })).toBeChecked();
};

export const clickEnergyDataset = async () => {
  await user.click(await screen.findByRole("checkbox", { name: "Energy [25]" }));
  expect(screen.getByRole("checkbox", { name: "Energy [25]" })).toBeChecked();
};

export const clickTransportDataset = async () => {
  await user.click(await screen.findByRole("checkbox", { name: "Transport [44]" }));
};

export const clickMedicalDataset = async () => {
  await user.click(await screen.findByRole("checkbox", { name: "Medical [32]" }));
};

export const expandPanel = async (panelName) =>
  await user.click(screen.getByRole("button", { name: panelName }));

export const AssetBtn = ({ label, assets, event, onElementClick }) => (
  <button
    onClick={() => {
      onElementClick(
        event,
        assets.find((asset) => asset.label === label)
      );
    }}
  >
    {label}
  </button>
);

export const CxnBtn = ({ label, connections, event, onElementClick }) => (
  <button
    onClick={() => {
      onElementClick(
        event,
        connections.find((cxn) => cxn.label === label)
      );
    }}
  >
    {label}
  </button>
);

export const PanelProviders = ({ children, elementsProviderValue }) => (
  <UseHttpProvider options={{ cacheLife: 0, cachePolicy: "no-cache" }}>
    <ElementsProvider value={elementsProviderValue}>{children}</ElementsProvider>
  </UseHttpProvider>
);

const LoadDataWrapper = ({ testComponent, children }) => (
  <MapProvider>
    <CytoscapeProvider>
      <ElementsProvider>
        <ElementsContext.Consumer>
          {({ assets, connections, selectedElements, onElementClick }) => {
            return (
              <>
                {testComponent && testComponent({ assets, connections, onElementClick })}
                <div id="all-assets">
                  {assets.map((asset) => (
                    <p id="asset" key={asset.id}>
                      {asset.id}
                    </p>
                  ))}
                </div>
                <div id="all-connections">
                  {connections.map((cxn) => (
                    <p id="cxn" key={cxn.id}>
                      {cxn.id}
                    </p>
                  ))}
                </div>
                <div id="selected-elements">
                  {selectedElements.map((selectedElement) => (
                    <p id="selected-element" key={selectedElement.id}>
                      {selectedElement.id}
                    </p>
                  ))}
                </div>
                <Dataset />
                {children}
              </>
            );
          }}
        </ElementsContext.Consumer>
      </ElementsProvider>
    </CytoscapeProvider>
  </MapProvider>
);

export const renderTestComponent = (ui, options) => {
  return {
    user: userEvent.setup(),
    ...render(ui, {
      wrapper: ({ children }) => (
        <LoadDataWrapper testComponent={options?.testComponent}>{children}</LoadDataWrapper>
      ),
      ...options?.testingLibraryOptions,
    }),
  };
};

const createTestQueryClient = () => {
  const queryClient = new QueryClient({
    defaultOptions: {
      queries: {
        retry: false,
      },
    },
  });
  return queryClient;
};
export const createQueryClientWrapper = () => {
  const queryClient = createTestQueryClient();
  return ({ children }) => (
    <QueryClientProvider client={queryClient}>{children}</QueryClientProvider>
  );
};

export const renderWithQueryClient = (ui, options) => {
  const queryClient = createTestQueryClient();

  // https://github.com/testing-library/react-testing-library/issues/218#issuecomment-436730757
  const rendered = render(
    <QueryClientProvider client={queryClient}>{ui}</QueryClientProvider>,
    options
  );

  return {
    user,
    ...rendered,
    rerender: (ui, options) =>
      renderWithQueryClient(ui, { container: rendered.container, ...options }),
  };
};

export const setup = (ui) => {
  return {
    user,
    ...render(ui),
  };
};

export const getCreatedAssets = async (
  assets,
  ids,
  findIcon = jest.fn(),
  getAssetGeometry = jest.fn()
) => {
  const createdAssets = (await createAssets(assets, findIcon, getAssetGeometry)).filter((asset) =>
    ids.includes(asset.id)
  );
  return createdAssets;
};

export const getCreatedDependencies = (dependencies, ids) => {
  const createdDependencies = createDependencies(dependencies).filter((asset) =>
    ids.includes(asset.id)
  );
  return createdDependencies;
};

export const DSProvidersWrapper = ({ children }) => (
  <DSProviders ontologyService={ontologyService}>{children}</DSProviders>
);
