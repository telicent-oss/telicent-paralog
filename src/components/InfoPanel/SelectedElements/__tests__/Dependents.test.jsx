import React from "react";
import { screen, waitForElementToBeRemoved, within } from "@testing-library/react";
import { http, HttpResponse, rest } from "msw";

import { ElementsProvider } from "context";
import { createParalogEndpoint } from "api/combined";
import {
  LARGE_WIND_FARM_ASSETS,
  COAL_PLANT_COMPLEX_ASSETS,
  server,
} from "mocks";
import { DSProvidersWrapper, getCreatedAssets, renderWithQueryClient } from "test-utils";
import { isAsset, isDependency } from "utils";

import Dependents from "../Dependents";

const renderAssetDependents = async ({ assets, element }) =>
  renderWithQueryClient(
    <DSProvidersWrapper>
      <ElementsProvider assets={assets}>
        <Dependents
          assetUri={element?.uri}
          dependent={element?.dependent}
          isAsset={isAsset(element)}
          isDependency={isDependency(element)}
        />
      </ElementsProvider>
    </DSProvidersWrapper>
  );

const renderE003AssetDetails = async (assets) => {
  const createdAssets = await getCreatedAssets(
    [
      ...LARGE_WIND_FARM_ASSETS,
      ...COAL_PLANT_COMPLEX_ASSETS,
    ],
    assets
  );
  const e003 = createdAssets.find((assets) => assets.id === "E003");
  return renderAssetDependents({ assets: createdAssets, element: e003 });
};

const waitForDependentsToLoad = async () => {
  await waitForElementToBeRemoved(() => screen.queryByText("Loading dependent assets"));
};

const toggleDependents = async (user) => {
  await user.click(screen.getByRole("button", { name: /dependent assets/i }));
};

describe("Dependents component", () => {
  beforeEach(() => {
    jest.clearAllMocks();
    jest.resetAllMocks();
  });

  test("does NOT render dependent assets when element is not defined", async () => {
    await renderE003AssetDetails([]);
    expect(screen.queryByTestId("dependent-assets")).not.toBeInTheDocument();
  });

  test("can toggle dependents assets", async () => {
    const { user } = await renderE003AssetDetails(["E003"]);
    await waitForDependentsToLoad();

    expect(
      screen.getByRole("heading", { name: "4 dependent assets", level: 3 })
    ).toBeInTheDocument();
    expect(screen.queryByRole("list")).not.toBeInTheDocument();

    await toggleDependents(user);
    expect(screen.getByRole("list")).toBeInTheDocument();

    const dependents = screen.getAllByRole("listitem");
    expect(dependents).toHaveLength(4);

    expect(
      within(dependents[0]).getByRole("heading", { name: "Jose 20kv Substation" })
    ).toBeInTheDocument();
    expect(within(dependents[0]).getByText("E008")).toBeInTheDocument();

    expect(
      within(dependents[3]).getByRole("heading", { name: "Best Coleman Power Station" })
    ).toBeInTheDocument();
    expect(within(dependents[3]).getByText("E001")).toBeInTheDocument();

    expect(dependents).toMatchSnapshot("dependents list");
  });

  test("does NOT show dependents when none are found", async () => {
    server.use(
      http.get(createParalogEndpoint("asset/dependents"), (req) => {
        const url = new URL(req.request.url);
        const assetUri = url.searchParams.get("assetUri");
        if (assetUri === "https://www.example.com/Instruments#E003") {
          return HttpResponse.json([], { status: 200 });
        }
      })
    );

    await renderE003AssetDetails(["E001", "E003"]);
    await waitForDependentsToLoad();

    expect(screen.queryByTestId("dependent-assets")).not.toBeInTheDocument();
  });

  test("renders error message when asset dependents cannot be retrieved", async () => {
    server.use(
      http.get(createParalogEndpoint("asset/dependents"), (req) => {

        const url = new URL(req.request.url);
        const assetUri = url.searchParams.get("assetUri");
        if (assetUri === "https://www.example.com/Instruments#E003") {
          return HttpResponse.json({
            message: "Depedents for https://www.example.com/Instruments#E003 not found",
          }, { status: 404 })
        }
      })
    );

    await renderE003AssetDetails(["E001", "E003"]);
    await waitForDependentsToLoad();

    expect(
      screen.getByText("Failed to retrieve dependents for https://www.example.com/Instruments#E003")
    ).toBeInTheDocument();
  });

  test("renders error message when one or more dependents returns an error", async () => {
    server.use(
      http.get(createParalogEndpoint("asset"), (req) => {

        const url = new URL(req.request.url);
        const assetUri = url.searchParams.get("assetUri");
        if (assetUri === "https://www.example.com/Instruments#E001") {
          return HttpResponse.json({
            message: "Asset information for https://www.example.com/Instruments#E001 not found",
          }, { status: 404 })
        }
      })
    );

    const { user } = await renderE003AssetDetails(["E001", "E003"]);
    await waitForDependentsToLoad();

    await toggleDependents(user);
    const dependents = screen.getAllByRole("listitem");
    expect(dependents).toHaveLength(4);

    expect(
      within(dependents[3]).getByText(
        "Failed to retrieve asset information for https://www.example.com/Instruments#E001"
      )
    ).toBeInTheDocument();
    expect(dependents).toMatchSnapshot("dependents with error(s)");
  });
});
