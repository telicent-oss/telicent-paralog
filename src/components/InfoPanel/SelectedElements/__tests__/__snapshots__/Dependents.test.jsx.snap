// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Dependents component > can toggle dependents assets > dependents list 1`] = `
[
  <li
    class="items-center p-2 rounded-md gap-x-2 bg-black-300"
  >
    <div
      class="flex items-center gap-x-2"
    >
      <div
        aria-label="SmallWindFarm"
        class="MuiAvatar-root MuiAvatar-circular MuiAvatar-colorDefault css-1bc90xp-MuiAvatar-root"
      >
        <i
          class="fa-solid fa-utility-pole"
          title="SmallWindFarm-icon"
        />
      </div>
      <div>
        <h4>
          Jose 20kv Substation
        </h4>
        <p
          class="text-sm uppercase"
        >
          small wind farm
        </p>
        <p
          class="text-sm"
        >
          E008
        </p>
      </div>
    </div>
    <p
      class="text-sm whitespace-nowrap"
    >
      Criticality: 
      6
    </p>
    <p
      class="text-sm whitespace-nowrap"
    >
      Connection Strength: 
      3
    </p>
  </li>,
  <li
    class="items-center p-2 rounded-md gap-x-2 bg-black-300"
  >
    <div
      class="flex items-center gap-x-2"
    >
      <div
        aria-label="SmallWindFarm"
        class="MuiAvatar-root MuiAvatar-circular MuiAvatar-colorDefault css-1bc90xp-MuiAvatar-root"
      >
        <i
          class="fa-solid fa-utility-pole"
          title="SmallWindFarm-icon"
        />
      </div>
      <div>
        <h4>
          West Coleman 50kV Substation
        </h4>
        <p
          class="text-sm uppercase"
        >
          small wind farm
        </p>
        <p
          class="text-sm"
        >
          E002
        </p>
      </div>
    </div>
    <p
      class="text-sm whitespace-nowrap"
    >
      Criticality: 
      8
    </p>
    <p
      class="text-sm whitespace-nowrap"
    >
      Connection Strength: 
      3
    </p>
  </li>,
  <li
    class="items-center p-2 rounded-md gap-x-2 bg-black-300"
  >
    <div
      class="flex items-center gap-x-2"
    >
      <div
        aria-label="SmallWindFarm"
        class="MuiAvatar-root MuiAvatar-circular MuiAvatar-colorDefault css-1bc90xp-MuiAvatar-root"
      >
        <i
          class="fa-solid fa-utility-pole"
          title="SmallWindFarm-icon"
        />
      </div>
      <div>
        <h4>
          Bluey 10kV Substation
        </h4>
        <p
          class="text-sm uppercase"
        >
          small wind farm
        </p>
        <p
          class="text-sm"
        >
          E012
        </p>
      </div>
    </div>
    <p
      class="text-sm whitespace-nowrap"
    >
      Criticality: 
      10
    </p>
    <p
      class="text-sm whitespace-nowrap"
    >
      Connection Strength: 
      3
    </p>
  </li>,
  <li
    class="items-center p-2 rounded-md gap-x-2 bg-black-300"
  >
    <div
      class="flex items-center gap-x-2"
    >
      <div
        aria-label="CoalPlantComplex"
        class="MuiAvatar-root MuiAvatar-circular MuiAvatar-colorDefault css-1bc90xp-MuiAvatar-root"
      >
        <i
          class="fa-regular fa-bolt-lightning"
          title="CoalPlantComplex-icon"
        />
      </div>
      <div>
        <h4>
          Best Coleman Power Station
        </h4>
        <p
          class="text-sm uppercase"
        >
          coal plant complex
        </p>
        <p
          class="text-sm"
        >
          E001
        </p>
      </div>
    </div>
    <p
      class="text-sm whitespace-nowrap"
    >
      Criticality: 
      2
    </p>
    <p
      class="text-sm whitespace-nowrap"
    >
      Connection Strength: 
      3
    </p>
  </li>,
]
`;

exports[`Dependents component > renders error message when one or more dependents returns an error > dependents with error(s) 1`] = `
[
  <li
    class="items-center p-2 rounded-md gap-x-2 bg-black-300"
  >
    <div
      class="flex items-center gap-x-2"
    >
      <div
        aria-label="SmallWindFarm"
        class="MuiAvatar-root MuiAvatar-circular MuiAvatar-colorDefault css-1bc90xp-MuiAvatar-root"
      >
        <i
          class="fa-solid fa-utility-pole"
          title="SmallWindFarm-icon"
        />
      </div>
      <div>
        <h4>
          Jose 20kv Substation
        </h4>
        <p
          class="text-sm uppercase"
        >
          small wind farm
        </p>
        <p
          class="text-sm"
        >
          E008
        </p>
      </div>
    </div>
    <p
      class="text-sm whitespace-nowrap"
    >
      Criticality: 
      6
    </p>
    <p
      class="text-sm whitespace-nowrap"
    >
      Connection Strength: 
      3
    </p>
  </li>,
  <li
    class="items-center p-2 rounded-md gap-x-2 bg-black-300"
  >
    <div
      class="flex items-center gap-x-2"
    >
      <div
        aria-label="SmallWindFarm"
        class="MuiAvatar-root MuiAvatar-circular MuiAvatar-colorDefault css-1bc90xp-MuiAvatar-root"
      >
        <i
          class="fa-solid fa-utility-pole"
          title="SmallWindFarm-icon"
        />
      </div>
      <div>
        <h4>
          West Coleman 50kV Substation
        </h4>
        <p
          class="text-sm uppercase"
        >
          small wind farm
        </p>
        <p
          class="text-sm"
        >
          E002
        </p>
      </div>
    </div>
    <p
      class="text-sm whitespace-nowrap"
    >
      Criticality: 
      8
    </p>
    <p
      class="text-sm whitespace-nowrap"
    >
      Connection Strength: 
      3
    </p>
  </li>,
  <li
    class="items-center p-2 rounded-md gap-x-2 bg-black-300"
  >
    <div
      class="flex items-center gap-x-2"
    >
      <div
        aria-label="SmallWindFarm"
        class="MuiAvatar-root MuiAvatar-circular MuiAvatar-colorDefault css-1bc90xp-MuiAvatar-root"
      >
        <i
          class="fa-solid fa-utility-pole"
          title="SmallWindFarm-icon"
        />
      </div>
      <div>
        <h4>
          Bluey 10kV Substation
        </h4>
        <p
          class="text-sm uppercase"
        >
          small wind farm
        </p>
        <p
          class="text-sm"
        >
          E012
        </p>
      </div>
    </div>
    <p
      class="text-sm whitespace-nowrap"
    >
      Criticality: 
      10
    </p>
    <p
      class="text-sm whitespace-nowrap"
    >
      Connection Strength: 
      3
    </p>
  </li>,
  <li
    class="p-2 bg-red-900 rounded-md"
  >
    Failed to retrieve asset information for https://www.example.com/Instruments#E001
  </li>,
]
`;
